File: problem3.js
/* testing number 600851475143 */
/* Solution: 6857 */

function greatestPrimeFactor(num) {
	
	function isPrime(n) {
		var prime = (n === 1) ? false : true;
		for (var i = 2; i <= Math.sqrt(n); i++) {
			if (n % i === 0) prime = false;
		}
		return prime;
	}
	
	var numArr = [];
	var lastNum = Math.sqrt(num);
	for (var j = 2; j <= lastNum; j++) {
		if (num % j === 0) {
			numArr.push(j);
			lastNum = num / j;
			numArr.push(lastNum);
		}
	}
	
	return numArr.filter(n => isPrime(n))
		.sort((a, b) => b - a)[0];
}

console.log(greatestPrimeFactor(600851475143));

File: problem_3.rb
# /* testing number 600851475143 */
# Solution: 6857

def greatest_prime(num)
  num_arr = []
  last_num = num**0.5

  (2...last_num).each do |j|
    next unless (num % j).zero?
    num_arr.push(j)
    last_num = num / j
    num_arr.push(last_num)
  end
  num_arr.select { |item| prime?(item) }.sort.last
end

def prime?(n)
  prime = n == 1 ? false : true

  (2...n**0.5).each { |i| prime = false if (n % i).zero? }
  prime
end

puts greatest_prime(600_851_475_143)
